File: Easy/21. Merge Two Sorted Lists/1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        def insertNode(l1 , l2):
            if l1 == None and l2 == None:
                return None
            l3 = ListNode()
            if l1 == None:
                l3.val = l2.val
                l3.next = insertNode(None , l2.next)
            elif l2 == None:
                l3.val = l1.val
                l3.next = insertNode(l1.next , None)
            elif l1.val <= l2.val:
                l3.val = l1.val
                l3.next = insertNode(l1.next , l2)
            else:
                l3.val = l2.val
                l3.next = insertNode(l1 , l2.next)
            return l3
        
        return insertNode(list1 , list2)

File: Medium/2. Add Two Numbers/1.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let dummyHead = new ListNode();
    let tail = dummyHead;
    let carry = 0;

    while(l1 !== null || l2 !== null || carry !== 0){
        let digit1 = (l1 !== null) ? l1.val : 0;
        let digit2 = (l2 !== null) ? l2.val : 0;

        let sum = digit1 + digit2 + carry;
        let digit = sum % 10
        carry = Math.floor(sum/10)

        let newNode = new ListNode(digit)
        tail.next = newNode
        tail = tail.next

        l1 = (l1 !== null) ? l1.next : null;
        l2 = (l2 !== null) ? l2.next : null;
    }
    let result = dummyHead.next
    dummyHead.next = null
    return result
};